[DevX][mergebot] Collapse failed rule into dropdown

PaliC opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94324

We create a new type of error `ExceptionWithInternalReason` which effectively does what `handle_exception` was doing before in creating an error, however, it lets you add extra internal input. 

We use this new error type in order to move the broken rule during merges to the  dev infra dropdown.

Nit: Also minor fix MandatoryChecksMissingError now actually uses the broken rule

Addresses: https://github.com/pytorch/test-infra/issues/1081

Result should be something in the vein of https://github.com/pytorch/pytorch/pull/93243#issuecomment-1421341230 (albeit with a different error message/summary)

PaliC committed:
[DevX][mergebot] Collapse failed rule into dropdown

[ghstack-poisoned]

PaliC committed:
Update on "[DevX][mergebot] Collapse failed rule into dropdown"




[ghstack-poisoned]

PaliC committed:
Update on "[DevX][mergebot] Collapse failed rule into dropdown"


We create a new type of error `MergeError` which effectively does what `handle_exception` was doing before in creating an error, however, it lets you add extra internal input. 

We use this new error type in order to move the broken rule during merges to the  dev infra dropdown.

Nit: Also minor fix MandatoryChecksMissingError now actually uses the broken rule

Addresses: https://github.com/pytorch/test-infra/issues/1081

Result should be something in the vein of https://github.com/pytorch/pytorch/pull/93243#issuecomment-1421341230 (albeit with a different error message/summary)


[ghstack-poisoned]

PaliC committed:
Update on "[DevX][mergebot] Collapse failed rule into dropdown"


We create a new type of error `MergeError` which effectively does what `handle_exception` was doing before in creating an error, however, it lets you add extra internal input. 

We use this new error type in order to move the broken rule during merges to the  dev infra dropdown.

Nit: Also minor fix MandatoryChecksMissingError now actually uses the broken rule

Addresses: https://github.com/pytorch/test-infra/issues/1081

Result should be something in the vein of https://github.com/pytorch/pytorch/pull/93243#issuecomment-1421341230 (albeit with a different error message/summary)


[ghstack-poisoned]

PaliC committed:
Update on "[DevX][mergebot] Collapse failed rule into dropdown"


We create a new type of error `MergeError` which effectively does what `handle_exception` was doing before in creating an error, however, it lets you add extra internal input. 

We use this new error type in order to move the broken rule during merges to the  dev infra dropdown.

Nit: Also minor fix MandatoryChecksMissingError now actually uses the broken rule

Addresses: https://github.com/pytorch/test-infra/issues/1081

Result should be something in the vein of https://github.com/pytorch/pytorch/pull/93243#issuecomment-1421341230 (albeit with a different error message/summary)


[ghstack-poisoned]

malfet commented on .github/scripts/trymerge.py:
```diff
@@ -1109,6 +1109,35 @@ def __init__(self, message: str, rule: Optional['MergeRule'] = None) -> None:
+            f"## {title}",
+            f"{exception}",
+            "",
+            f"{internal_debugging_formatted}"
```
`internal_debugging_formatted` would be None if `len(internal_debugging) == 0` wouldn't it?

malfet commented on .github/scripts/trymerge.py:
```diff
@@ -1109,6 +1109,35 @@ def __init__(self, message: str, rule: Optional['MergeRule'] = None) -> None:
+        exception = f"**Reason**: {reason}"
+
+        run_url = os.getenv("GH_RUN_URL")
+        internal_debugging = ""
+        if internal_info is not None:
+            internal_debugging += internal_info + "\n\n"
+        if run_url is not None:
+            internal_debugging += f"Raised by <a href=\"{run_url}\">workflow job</a>"
+
+        if len(internal_debugging) > 0:
+            # Hide this behind a collapsed bullet since it's not helpful to most devs
+            internal_debugging_formatted = "\n".join((
+                "<details><summary>Details for Dev Infra team</summary>",
+                "",
+                f"{internal_debugging}"
+                "",
+                "</details>"
+            ))
```
Define `internal_debugging` as list of strings and append/extend it. This way you can do single join operation at the very end
```suggestion
        internal_debugging: List[str] = []
        if internal_info is not None:
            internal_debugging.extend([internal_info, "", ""])
        if run_url is not None:
            internal_debugging.append( f"Raised by <a href=\"{run_url}\">workflow job</a>")

        if len(internal_debugging) > 0:
            # Hide this behind a collapsed bullet since it's not helpful to most devs
            internal_debugging= ["<details><summary>Details for Dev Infra team</summary>", ""] + internal_debugging + ["", "</details>"]
```

malfet commented on .github/scripts/trymerge.py:
```diff
@@ -1109,6 +1109,35 @@ def __init__(self, message: str, rule: Optional['MergeRule'] = None) -> None:
         self.rule = rule
 
+class MergeError(Exception):
+    def __init__(self, reason: Union[str, Exception], internal_info: Optional[str] = None, title: str = "Merge failed"):
```
```suggestion
    def __init__(self, reason: Union[str, Exception], internal_info: Optional[str] = None, title: str = "Merge failed") -> None:
```

malfet commented on .github/scripts/trymerge.py:
```diff
@@ -1109,6 +1109,35 @@ def __init__(self, message: str, rule: Optional['MergeRule'] = None) -> None:
         super().__init__(message)
         self.rule = rule
 
+class MergeError(Exception):
```
Hmm, it's not really a `MergeError`, but rather `ExceptionWithReason`? And shouldn't it be subclassed from `RuntimeError` rather than `Exception`?

PaliC replied:
The error ends up capturing things which are not RuntimeErrors (as it just replaces `handleException`) like MandatoryChecksMissingError, so probably best to subclass to Exception. Though I'll change the name to make it more clear.

malfet reviewed (changes requested):
This would always add `None` when `internal_debugging_info` is missing.

The rest are nits

PaliC committed:
Update on "[DevX][mergebot] Collapse failed rule into dropdown"


We create a new type of error `MergeError` which effectively does what `handle_exception` was doing before in creating an error, however, it lets you add extra internal input. 

We use this new error type in order to move the broken rule during merges to the  dev infra dropdown.

Nit: Also minor fix MandatoryChecksMissingError now actually uses the broken rule

Addresses: https://github.com/pytorch/test-infra/issues/1081

Result should be something in the vein of https://github.com/pytorch/pytorch/pull/93243#issuecomment-1421341230 (albeit with a different error message/summary)


[ghstack-poisoned]

PaliC committed:
Update on "[DevX][mergebot] Collapse failed rule into dropdown"


We create a new type of error `MergeError` which effectively does what `handle_exception` was doing before in creating an error, however, it lets you add extra internal input. 

We use this new error type in order to move the broken rule during merges to the  dev infra dropdown.

Nit: Also minor fix MandatoryChecksMissingError now actually uses the broken rule

Addresses: https://github.com/pytorch/test-infra/issues/1081

Result should be something in the vein of https://github.com/pytorch/pytorch/pull/93243#issuecomment-1421341230 (albeit with a different error message/summary)


[ghstack-poisoned]
